common/src/main/java/com/google/auto/common/BasicAnnotationProcessor.java

```diff
@@ -511,73 +521,372 @@ private ImmutableSetMultimap<Class<? extends Annotation>, Element> toClassKeyedM
+       * Overridden methods.
+       * 2. As for Overloaded executable elements, according to
+       * 2.1. JLS 8.4.2:
+       * "The signature of a method m1 is a subsignature of the signature of a method m2 if either:
+       * * m2 has the same signature as m1, or
+       * * the signature of m1 is the same as the erasure (§4.6) of the signature of m2."
+       * Therefore, an ordered erased parameters types is stored.
+       * 2.2. JLS 8.4.9:
+       * "There is no required relationship between the return types or between the throws clauses
+       * of two methods with the same name, unless their signatures are override-equivalent."
+       * Therefore, checking for simple name and erased parameter types is enough to identify the
+       * represented executable element.
+       */
+      this.simpleName = element.getSimpleName();
+      this.typeUtils = Objects.requireNonNull(typeUtils);
+      this.explicitOverrides = new ExplicitOverrides(typeUtils);
+      this.erasedParametersTypes =
+          // It might return null, if the method is not found in enclosingTypeElement
+          Objects.requireNonNull(
+              explicitOverrides.erasedParameterTypes(execElement, enclosingTypeElement));
+    }
+
+    @Override
+    protected boolean isAcceptableElementKind(Element element) {
+      return element.getKind() == ElementKind.CONSTRUCTOR
+          || element.getKind() == ElementKind.METHOD;
+    }
+
+    /* Executable element can be identified using the enclosing type element,
+     * simple name, and an ordered erased parameter types as described in the
+     * constructor.
      */
-    Optional<? extends Element> getElement(Elements elements) {
-      return Optional.fromNullable(
-          kind == Kind.PACKAGE_NAME
-              ? elements.getPackageElement(name)
-              : elements.getTypeElement(name));
+    @Override
+    Optional<ExecutableElement> getElement(Elements eltUtils) {
+      Optional<TypeElement> optionalEnclosingTypeElement =
+          enclosingTypeElementName.getElement(eltUtils);
+      if (!optionalEnclosingTypeElement.isPresent()) {
+        return Optional.absent();
+      }
+      TypeElement enclosingTypeElement = optionalEnclosingTypeElement.get();
+
+      return Optional.of(
+          (ExecutableElement)
+              enclosingTypeElement.getEnclosedElements().stream()
+                  .filter(
+                      element ->
+                          isAcceptableElementKind(element)
+                              && simpleName.equals(element.getSimpleName())
+                              && hasSameErasedParametersType(
+                                  Objects.requireNonNull(
+                                      explicitOverrides.erasedParameterTypes(
+                                          (ExecutableElement) element, enclosingTypeElement))))
+                  .collect(MoreCollectors.onlyElement()));
+    }
+
+    private boolean hasSameErasedParametersType(
```

Comment:
I realize that the fix I suggested isn't really enough. We would also have to handle the case where we have `Set<BarFactory>`, `Set<? extends BarFactory>`, even a `<T extends BarFactory>` type parameter on the method. That would suggest writing a `TypeVisitor` similar to [the one](https://github.com/google/auto/blob/01bf873bba2b23e6c07d9451da4fc4f832b87a43/common/src/main/java/com/google/auto/common/MoreTypes.java#L190) used by `MoreTypes.equivalence()`.

However!

I think there is an easier way. What we're trying to do is to find this method again in a later round after we saw it in an earlier one. And the issue is that there might be other methods with the same name but different parameter types. But we should be able to rely on the specification of [`TypeElement.getEnclosedElements()`](https://docs.oracle.com/en/java/javase/17/docs/api/java.compiler/javax/lang/model/element/TypeElement.html#getEnclosedElements()), which essentially says that the elements will be returned in source order. That order shouldn't change between rounds, so you can just remember that this is the second method called `foo`, or even just the second element overall, to find it again. (That might fail if something like Lombok is hacking the AST, but I think that's an acceptable restriction.)

I'm going to see if I can make this idea work and run it past all of Google's compilations again.

Reply:
Actually, we're looking at erased types, aren't we? So none of the cases I mentioned applies, though arrays do. Nevertheless I think this idea is probably cleaner than comparing the `getSimpleName()` of a `TypeElement` against the `getSimpleName()` of `typeUtils.asElement(errorType)`.

Reply:
Nicely spotted, thank you!

I have implemented this lookup using your suggestion.

I need your opinion regarding one slight problem that I have found.
Interestingly, I found that `Types#isSameType()` returns true for the `ERROR` type regardless of the other type. In hindsight this makes sense. One thing I have noticed about (open-)JDK behavior is the following which is peculiar and is a slight problem for our efforts. I believe this slight problem is the result of the aforementioned fact. Let me explain using an example:
``` java
@Annotation
<C extends Set<String>> void overloadedMethod(C x) {}
@Annotation
<C extends SortedSet<String>> void overloadedMethod(C x) {}
@Annotation
void overloadedMethod(SomeToBeGeneratedClass x) {}
@Annotation
<T extends TreeSet<String>> void overloadedMethod(T x) {}
```

When javac is scanning for annotated elements, and `SomeToBeGeneratedClass` references an `ERROR` type, then the last overloaded method is not reported by `RoundEnvironment#getElementsAnnotatedWith()`. Consequently, it will never get processed. It may be beneficial to note that the first two methods, although having the same `toString`,  are reported. Moreover, if the last method position moves above the `ERROR` referencing method, then in that case it will be reported. This feels like a bug to me, and if indeed it is a bug it would be great if I be able to contribute to open-JDK and fix it. Do you think this is an open-JDK bug? In the meantime, I believe this problem can only occur for scenarios exactly like this one. I can put simple check for it and simply handle it. Please let me know what you think. I will push this simple check in the near future.

Reply:
I did not modify the implementation of deferring the ill-informed elements sufficiently before. After noticing it, every problem is solved. I believe now the code is even more readable, I hope you find it the same.